src/Post/Post.php

```diff
@@ -104,6 +104,36 @@ public static function boot()
         static::addGlobalScope(new RegisteredTypesScope);
     }
 
+    /**
+     * We override this so that we can use custom logic to.
+     */
+    protected function insertAndSetId(Builder $query, $attributes)
+    {
+        $attrsNoNumber = array_filter($attributes, function ($key) {
+            return $key !== 'number';
+        }, ARRAY_FILTER_USE_KEY);
+
+        $literalsSubquery = Post::query();
+
+        foreach ($attrsNoNumber as $attr => $value) {
+            //! DANGER!!! Literal Injection Vuln!!!
+            $literalsSubquery->selectRaw("'$value' as $attr");
+        }
+
+        $literalsSubquery->selectSub(Post::query()->where('discussion_id', $this->discussion_id)->selectRaw('count(*)+1'), 'number');
+
+        $attrNames = array_keys($attrsNoNumber);
+        $attrNames[] = 'number';
+
+        $query->insertUsing($attrNames, $literalsSubquery->limit(1));
+
+        /** @var MySqlConnection */
+        $con = $query->getQuery()->getConnection();
+        $id = $con->getPdo()->lastInsertId('id');
```

Comment:
Some things:

Use the `getKeyName` here again.

`lastInsertId` allows an argument, the argument needs to relate to "the name of a sequence object for the name parameter", not a column I think?

I looked up whether using lastInsertId is safe on highly volatile communities, it seems this query is specific to each connection and so it will return the right value.

Last insert Id might not be the last inserted one on high volatile communities.

Reply:
It should be safe, that's how the current default implementation works.